File: polyformalism-languages/conservation-law/conservation.c
/* C: What hardware truth reveals about conservation
   Forces you to think about: Hardware. int is 32 bits. double is IEEE 754.
     sqrt() calls a hardware instruction. Mathematics is EXECUTED ON PHYSICAL HARDWARE.
   Broken assumption: That abstraction is free. C has no generics, no closures.
   Novel insight: The conservation law is a PHYSICAL PROCESS. When the CPU computes
     δ(n), electrons move through silicon. The computation has a COST — joules,
     nanoseconds. Every δ(n) costs energy, and the conservation law is about energy.

   Build: cc conservation.c -lm -o conservation && ./conservation
*/

#include <stdio.h>
#include <math.h>

double delta(int n) {
    return (1.0 / sqrt((double)n)) * (1.0 - 3.0 / (2.0 * n));
}

int main(void) {
    printf("%5s  %10s  %8s\n", "n", "δ(n)", "CCR%");
    printf("%.28s\n", "----------------------------");
    for (int n = 1; n <= 100; n++) {
        printf("%5d  %10.6f  %7.2f%%\n", n, delta(n), delta(n) * 100.0);
    }
    return 0;
}

File: polyformalism-languages/conservation-law/conservation.pl
% PROLOG: What declarative logic reveals about conservation
% Forces you to think about: Relations, not functions. delta(N, D) is not
%   "compute D from N" — it's "N and D stand in this RELATION."
% Broken assumption: That computation flows input→output. Given η and C,
%   you can query γ. The conservation law is BIDIRECTIONALLY QUERYABLE.
% Novel insight: γ + η = C is a Prolog clause. Directionality is an assumption,
%   not a law. You can ask "for what n is δ(n) > 0.1?" and let Prolog search.

:- use_module(library(lists)).

delta(N, D) :-
    N > 0,
    D is (1 / sqrt(N)) * (1 - 3 / (2 * N)).

ccr(N, C) :-
    delta(N, D),
    C is D * 100.

conservation_table :-
    writeln('    n       δ(n)     CCR%'),
    writeln('----------------------------'),
    forall(between(1, 100, N),
           ( delta(N, D),
             ccr(N, C),
             format('~5d  ~10f  ~7f%~n', [N, D, C]) )).

% Query: ?- conservation_table.
% Query: ?- delta(N, D), D > 0.1.  (find all n where δ > 0.1)

